fs-cache/src/cache.rs

```diff
@@ -0,0 +1,668 @@
+    fn persist_to_disk(&mut self, key: &K, value: &V) -> Result<()> {
+        log::debug!("cache/{}: writing to disk for key {}", self.label, key);
+        fs::create_dir_all(&self.path)?;
+
+        let file_path = self.path.join(key.to_string());
+        debug_assert!(
+            !file_path.exists(),
+            "File {} should not exist before writing",
+            file_path.display()
+        );
+
+        temp_and_move(value.as_ref(), self.path.clone(), &key.to_string())
+            .map_err(|err| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    format!("Failed to write value for key {}: {}", key, err),
+                )
+            })
+    }
+
+    /// Reads a value from disk.
+    fn read_from_disk(&self, key: &K) -> Result<V>
+    where
+        V: From<Vec<u8>>, // Add trait bound for reading binary data
+    {
+        let file_path = self.path.join(key.to_string());
+        let contents = fs::read(&file_path)?;
+        Ok(V::from(contents))
+    }
+
+    /// Returns the size of a value in bytes.
+    ///
+    /// First checks the memory cache for size information to avoid disk access.
+    /// Falls back to checking the file size on disk if not found in memory.
+    fn get_file_size(&self, key: &K) -> Result<usize> {
+        Ok(fs::metadata(self.path.join(key.to_string()))?.len() as usize)
+    }
+
+    /// Adds or updates a value in the memory cache, evicting old entries if needed.
+    /// Returns error if value is larger than maximum memory limit.
+    fn update_memory_cache(&mut self, key: &K, value: &V) -> Result<()> {
+        log::debug!("cache/{}: caching in memory for key {}", self.label, key);
+        let size = self.get_file_size(key)?;
+
+        // If single value is larger than total limit, just skip memory caching
+        if size > self.max_memory_bytes {
+            log::error!(
+                "cache/{}: value size {} exceeds limit {}",
+                self.label,
+                size,
+                self.max_memory_bytes
+            );
+            return Ok(());
+        }
+
+        // Remove oldest entries until we have space for new value
+        while self.current_memory_bytes + size > self.max_memory_bytes {
+            let (_, old_entry) = self
+                .memory_cache
+                .pop_lru()
+                .expect("Cache should have entries to evict");
+            debug_assert!(
+                self.current_memory_bytes >= old_entry.size,
+                "Memory tracking inconsistency detected"
+            );
+            self.current_memory_bytes = self
+                .current_memory_bytes
+                .saturating_sub(old_entry.size);
+        }
+
+        // Add new value and update size
+        self.memory_cache.put(
+            key.clone(),
+            CacheEntry {
+                value: value.clone(),
+                size,
+            },
+        );
```

Comment:
A small suggestion: we could use `push` here instead of `put`. According to the docs, `push` does the following: ‘If an entry with key k already exists in the cache or another cache entry is removed (due to the LRU’s capacity), then it returns the old entry’s key-value pair.’

It should be a nice touch to log the evicted key, if there is one.

Reply:
Added!